File: goombah/m/tests/compare_goombah_and_pounders.m
% This wrapper tests various algorithms against the Benchmark functions from the
% More and Wild SIOPT paper "Benchmarking derivative-free optimization algorithms"
function [] = compare_goombah_and_pounders()
addpath('../../../../BenDFO/data/');
addpath('../../../../BenDFO/m/');
addpath('../../../minq/m/minq5');
addpath('../../../goombah/m');
addpath('../../../pounders/m/general_h_funs');
addpath('../../../pounders/m');
addpath('../../../goombah/m/subproblems/');
addpath('../../../manifold_sampling/m');
addpath('../../../manifold_sampling/m/h_examples/');

load dfo.dat;

spsolver = 2; % TRSP Solver
nfmax = 100;
gtol = 1e-13;
factor = 10;

for row = 1:length(dfo)
    nprob = dfo(row, 1);
    n = dfo(row, 2);
    m = dfo(row, 3);
    factor_power = dfo(row, 4);

    BenDFO.nprob = nprob;
    BenDFO.m = m;
    BenDFO.n = n;

    objective = @(x)calfun_wrapper(x, BenDFO, 'smooth');

    X0 = dfoxs(n, nprob, factor^factor_power)';
    npmax = 2 * n + 1;  % Maximum number of interpolation points [2*n+1]
    L = -Inf(1, n);
    U = Inf(1, n);
    nfs = 0;
    F0 = [];
    xkin = 1;
    delta = 0.1;
    printf = 0;

    GAMS_options.file = '../../../goombah/m/subproblems/minimize_sum_quad_models_squared.gms';
    GAMS_options.solvers = 1:4;
    subprob_switch = 'linprog';

    for hfun_cases = 1:1
        Results = cell(2, 3, 53);
        if hfun_cases == 1
            hfun = @(F)sum(F.^2);
            combinemodels = @leastsquares;
        elseif hfun_cases == 2
            error("Not implemented in GOOMBAH yet");
            alpha = 0; % If changed here, also needs to be adjusted in squared_diff_from_mean.m
            hfun = @(F)sum((F - 1 / length(F) * sum(F)).^2) - alpha * (1 / length(F) * sum(F))^2;
            combinemodels = @squared_diff_from_mean;
        elseif hfun_cases == 3
            error("Not implemented in GOOMBAH yet");
            if m ~= 3 % Emittance is only defined for the case when m == 3
                continue
            end
            hfun = @emittance_h;
            combinemodels = @emittance_combine;
        end
        disp([row, hfun_cases]);

        filename = ['./benchmark_results/poundersM_and_GOOMBAH_nfmax=' int2str(nfmax) '_gtol=' num2str(gtol) '_prob=' int2str(row) '_spsolver=' num2str(spsolver) '_hfun=' func2str(combinemodels) '.mat'];

        for method = [1, 2]
            if method == 1
                [X, F, flag, xk_best] = pounders(objective, X0, n, npmax, nfmax, gtol, delta, nfs, m, F0, xkin, L, U, printf, spsolver, hfun, combinemodels);
                Results{method, hfun_cases, row}.alg = 'POUNDERs';
            elseif method == 2
                [X, F, h, xkin] = goombah(@sum_squared, objective, nfmax, X0, L, U, GAMS_options, subprob_switch);
                Results{method, hfun_cases, row}.alg = 'GOOMBAH';
            end

            evals = size(F, 1);
            h = zeros(evals, 1);
            for i = 1:evals
                h(i) = hfun(F(i, :));
            end
        end

        Results{method, hfun_cases, row}.problem = ['problem ' num2str(row) ' from More/Wild'];
        Results{method, hfun_cases, row}.Fvec = F;
        Results{method, hfun_cases, row}.H = h;
        Results{method, hfun_cases, row}.X = X;
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        %     save('-mat7-binary', filename, 'Results') % Octave save
        save(filename, 'Results');
    end
end
end

function [fvec] = calfun_wrapper(x, struct, probtype)
[~, fvec, ~] = calfun(x, struct, probtype);
end

File: goombah/py/tests/compare_goombah_and_pounders.py
import os
import sys

import numpy as np
import scipy as sp

sys.path.append("../../../../minq/py/minq5/")  # Needed for spsolver=2
sys.path.append("../")  
from ibcdfo.pounders import pounders

BenDFO_root = "../../../../BenDFO/"
sys.path.append(BenDFO_root + "py/")  # Needed for spsolver=2
sys.path.append('../../../manifold_sampling/py/')
sys.path.append('../../../pounders/py/')
sys.path.append('../../../manifold_sampling/py/h_examples')

from sum_squared import sum_squared
from dfoxs import dfoxs
from calfun import calfun
from goombah import goombah

os.makedirs("benchmark_results", exist_ok=True)

dfo = np.loadtxt(BenDFO_root + "data/dfo.dat")

spsolver = 2  # TRSP solver
nfmax = 50
gtol = 1e-13
factor = 10

Results = {}
GAMS_options = {}
hfun = lambda F: np.sum(F**2)

for row, (nprob, n, m, factor_power) in enumerate(dfo):
    n = int(n)
    m = int(m)

    def objective(y):
        # It is possible to have python use the same objective values via
        # octave. This can be slow on some systems. To (for example)
        # test difference between matlab and python, used the following
        # line and add "from oct2py import octave" on a system with octave
        # installed.
        # out = octave.feval("calfun_wrapper", y, m, nprob, "smooth", [], 1, 1)
        out = calfun(y, m, int(nprob), "smooth", 0, vecout=True)
        assert len(out) == m, "Incorrect output dimension"
        return np.squeeze(out)

    X0 = dfoxs(n, nprob, int(factor**factor_power))
    npmax = 2 * n + 1  # Maximum number of interpolation points [2*n+1]
    L = -np.inf * np.ones((1, n))  # 1-by-n Vector of lower bounds [zeros(1, n)]
    U = np.inf * np.ones((1, n))  # 1-by-n Vector of upper bounds [ones(1, n)]
    nfs = 1
    F0 = np.zeros((1, m))
    F0[0] = objective(X0)
    xind = 0
    delta = 0.1
    printf = False

    filename = "./benchmark_results/poundersM_and_GOOMBAH_nfmax=" + str(nfmax) + "_gtol=" + str(gtol) + "_prob=" + str(row) + "_spsolver=" + str(spsolver) + ".mat"

    Results[row] = {}
    for method in [1,2]:
        if method == 1:
            [X, F, flag, xk_best] = pounders(objective, X0, n, npmax, nfmax, gtol, delta, nfs, m, F0, xind, L, U, printf, spsolver)
            Results[row]["alg"] = "pounders4py"
        elif method == 2:
            GAMS_options["solvers"] = range(4)
            [X, F, flag, xk_best] = goombah(sum_squared, objective, nfmax, X0, L, U, GAMS_options, "linprog")
            Results[row]["alg"] = "goombah"

        evals = F.shape[0]
        h = np.zeros(evals)

        for i in range(evals):
            h[i] = hfun(F[i, :])

        Results[row]["problem"] = "problem " + str(row) + " from More/Wild"
        Results[row]["Fvec"] = F
        Results[row]["H"] = h
        Results[row]["X"] = X


np.save("Results.npy", Results)
